libr/bin/format/pyc/pyc_magic.c

```diff
@@ -1,282 +1,658 @@
+	case 0x0a0d0de8: // 3560
+		return (struct pyc_version){ 0x0a0d0de8, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0de9: // 3561
+		return (struct pyc_version){ 0x0a0d0de9, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0dea: // 3562
+		return (struct pyc_version){ 0x0a0d0dea, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0deb: // 3563
+		return (struct pyc_version){ 0x0a0d0deb, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0dec: // 3564
+		return (struct pyc_version){ 0x0a0d0dec, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0ded: // 3565
+		return (struct pyc_version){ 0x0a0d0ded, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0dee: // 3566
+		return (struct pyc_version){ 0x0a0d0dee, "3.13a1", "Unkonwn commit" };
+	case 0x0a0d0e10: // 3600
+		// 3.14 will have 3600, will it be 3.14a1? IDK, update below if do Mr. time traveler.
+		return (struct pyc_version){ 0x0a0d0e10, "3.14", "Unkonwn commit" };
+	default:
+		return (struct pyc_version){ -1, NULL, NULL };
+	}
+}
+
+// in order of presidence, ie 3.0(END) > 3.0a1
+enum {
+	INVALID = -1,
+	ALPHA,
+	BETA,
+	RC,
+	END,
+	DOT
 };
 
-struct pyc_version get_pyc_version(ut32 magic) {
-	struct pyc_version fail = { -1, 0, 0 };
-	size_t i;
-	for (i = 0; i < sizeof (versions) / sizeof (*versions); i++) {
-		if (versions[i].magic == magic) {
-			return versions[i];
+static inline int septype (char **s, bool *err) {
+	switch (**s) {
+	case '.':
+		(*s)++;
+		char *bookmark;
+		if (!strtol (*s, &bookmark, 10)) {
+			// .0 is special b/c version.parse('3.0a1') == version.parse('3.000000a1') == version.parse('3a1') == version.parse('3.a1')
+			// So is this .0 isnignificant? Not if it's the .0 in 3.0.1, since 3.0.1 != 3.1
+			int ret = septype (&bookmark, err);
+			if (*err || ret != DOT) {
+				*s = bookmark;
+				return ret;
+			}
+		}
+		return DOT;
+	case 'a':
+		(*s)++;
+		return ALPHA;
+	case 'b':
+		(*s)++;
+		return BETA;
+	case 'r':
+		(*s)++;
+		if (**s != 'c') {
+			break;
 		}
+		// fallthrough
+	case 'c':
+		(*s)++;
+		return RC;
+	case '\0':
+		return END;
 	}
-	return fail;
+	*err = true;
+	return INVALID;
 }
 
-bool magic_int_within(ut32 target_magic, ut32 lower, ut32 upper, bool *error) {
-	if (*error) {
-		return false;
+static inline int py_vint_diff (char **va, char **vb, bool *err) {
+	char *holda, *holdb;
+	long ma = strtol (*va, &holda, 10);
+	long mb = strtol (*vb, &holdb, 10);
+	if (*va == holda || *vb == holdb) {
+		*err = true;
+		return -1;
 	}
-	ut64 ti = 0, li = 0, ui = 0;
-	ut64 size = sizeof (versions) / sizeof (struct pyc_version);
-	for (; ti < size && versions[ti].magic != target_magic; ti++) {
-		// just loop
+
+	*va = holda;
+	*vb = holdb;
+	return ma - mb;
+}
+
+/*
+ * Bassed on PEP 440 (https://peps.python.org/pep-0440/)
+ * This algorythm is good enough for the versions we have. We shouldn't need to
+ * work with `dev` or `+` versions since we only care about versions with
+ * unique pyc magic.
+ */
+int py_version_cmp(char *va, char *vb, bool *err) {
```

Comment:
arent va/vb const ?

Reply:
dunno why th elinter complains about this line :?

Reply:
git grep -n 'cmp(' libr | grep -v R_API | grep -v static | grep c:) && exit 1 I think thats the issue from lint.sh